crates/ruff/src/rules/refurb/rules/reimplemented_starmap.rs

```diff
@@ -0,0 +1,304 @@
+/// passed_all_tests = all(starmap(passed_test, zip(scores, passing_scores)))
+/// ```
+///
+/// ## References
+/// - [Python documentation: `itertools.starmap`](https://docs.python.org/3/library/itertools.html#itertools.starmap)
+#[violation]
+pub struct ReimplementedStarmap;
+
+impl Violation for ReimplementedStarmap {
+    const AUTOFIX: AutofixKind = AutofixKind::Sometimes;
+
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        format!("Use `itertools.starmap` instead of the generator")
+    }
+
+    fn autofix_title(&self) -> Option<String> {
+        Some(format!("Replace with `itertools.starmap`"))
+    }
+}
+
+/// A abstract node that can be considered a candidate for replacement with `starmap`.
+pub(crate) trait StarmapEquivalent {
+    /// Get generated element.
+    fn get_element(&self) -> &Expr;
+    /// Get generator comprehensions.
+    fn get_generators(&self) -> &[ast::Comprehension];
+    /// Get fix suggestion transforming this node into a call to `starmap`.
+    fn make_suggestion(name: String, iter: &Expr, func: &Expr, generator: Generator) -> String;
+    /// Check if it is possible to apply the fix in the current context.
+    fn check_fix_possibility(semantic: &SemanticModel) -> bool;
+}
+
+// FURB140
+pub(crate) fn reimplemented_starmap<T: StarmapEquivalent + Ranged>(
+    checker: &mut Checker,
+    generator: &T,
+) {
+    // Generator should have exactly one comprehension.
+    let [comprehension @ ast::Comprehension { .. }] = generator.get_generators() else {
+        return;
+    };
+
+    // This comprehension should have a form:
+    // ```python
+    // (x, y, z, ...) in iter
+    // ```
+    //
+    // `x, y, z, ...` are what we call `elts` for short.
+    let Some((elts, iter)) = match_comprehension(comprehension) else {
+        return;
+    };
+
+    // Generator should produce one element that should look like:
+    // ```python
+    // func(a, b, c, ...)
+    // ```
+    //
+    // here we refer to `a, b, c, ...` as `args`.
+    //
+    // NOTE: `func` is not necessarily just a function name, it can be an attribute access,
+    //       or even a call itself.
+    let Some((args, func)) = match_call(generator.get_element()) else {
+        return;
+    };
+
+    // Here we want to check that `args` and `elts` are the same (same length, same elements,
+    // same order).
+    if elts.len() != args.len()
+        || !std::iter::zip(elts, args)
+            // We intentionally do not use ComparableExpr here because it will compare expression
+            // contexts and in `elts` it's definitely `Load`, while in `args` it's always `Store`.
+            //
+            // For this reason, we compare names directly.
+            .all(|(x, y)| get_name_id(x) == get_name_id(y))
+    {
+        return;
+    }
+
+    let mut diagnostic = Diagnostic::new(ReimplementedStarmap {}, generator.range());
+    if checker.patch(diagnostic.kind.rule()) && T::check_fix_possibility(checker.semantic()) {
```

Comment:
Nit: Could `check_fix_possibility` and `make_suggestion` be combined so that the enum has a method `try_fix` which returns an `Option` instead?

Reply:
Yeah, I see what you mean. Will do!